### PR TITLE
feat: optimize API responses for LLM context efficiency

### LINKED ISSUE #84: feat: Optimize API responses for LLM context efficiency
## 🎯 Goal

Reduce LLM context window waste by filtering, formatting, and summarizing API responses before returning them to the LLM.

**Definition of Done:**

- [ ] ResponseFormatter strips non-essential fields from all API responses
- [ ] List endpoints include automatic summary statistics
- [ ] `compact` mode returns summaries without individual records
- [ ] Response size reduced by 40%+ for typical list queries

### Why This Matters

Current tools return raw freee API responses to the LLM, including:
- Internal IDs that aren't useful for analysis
- Deeply nested metadata (update timestamps, creator info, etc.)
- Verbose field names and structures
- No summary statistics on list results

This wastes LLM context window on data that adds no analytical value. A typical `freee_get_deals` response for 100 deals can be 50KB+ of JSON.

### Current State

- Raw freee API JSON returned directly to LLM
- No field filtering or formatting
- No summary statistics on list results
- LLM context window filled with low-value data

### Desired State

- Responses contain only business-meaningful data
- List responses include automatic summary statistics
- Compact mode available for when only totals are needed
- 40%+ reduction in response size

### User Stories

- As an LLM, I want concise responses so that I have more context window for reasoning
- As an LLM user, I want summary statistics on lists so that quick questions don't require processing 100 records
- As an LLM user, I want a compact mode so that overview queries are fast and token-efficient

### Functional Requirements

1. **ResponseFormatter class** — Strip non-essential fields from API responses
   ```typescript
   class ResponseFormatter {
     static formatDeal(deal: any): FormattedDeal {
       return {
         id: deal.id,
         issueDate: deal.issue_date,
         type: deal.type,
         amount: deal.amount,
         partner: deal.partner?.name,
         status: deal.status,
         details: deal.details?.map((d: any) => ({
           accountItem: d.account_item_name,
           amount: d.amount,
           taxCode: d.tax_code,
           description: d.description,
         })),
       };
     }

     static formatInvoice(invoice: any): FormattedInvoice { ... }
     static formatPartner(partner: any): FormattedPartner { ... }
     // ... formatters for each entity type
   }
   ```

   **Fields to keep**: business-meaningful data (names, amounts, dates, statuses)
   **Fields to strip**: internal IDs (unless needed for cross-referencing), audit metadata, empty arrays, null values

2. **Summary statistics on list responses**
   ```typescript
   // Before: just raw array
   { deals: [...100 items...] }

   // After: array + summary
   { summary: { totalCount: 100, hasMore: true, totalIncome: 5000000, totalExpense: 3000000 },
     deals: [...100 items (formatted)...] }
   ```

3. **Compact mode parameter**
   - Add optional `compact: boolean` parameter to list tools
   - When true: return summary statistics only (no individual records)
   - Useful when LLM just needs totals, not individual items
   - Example compact response: `{ summary: { totalCount: 247, totalIncome: 15000000, totalExpense: 8500000, dateRange: '2025-04-01 to 2025-12-31' } }`

### Technical Constraints

- **Backward compatibility**: Existing response structure should not break LLM workflows
- **ID preservation**: Keep IDs that are needed for cross-referencing (e.g., deal ID for follow-up queries)
- **Null handling**: Strip null/undefined/empty values to reduce JSON size

### Functional

- [ ] `ResponseFormatter` class created with formatters for each entity type (deals, invoices, partners, account items, etc.)
- [ ] List endpoints return summary statistics alongside results
- [ ] `compact` parameter supported on list endpoints (deals, invoices, partners, account items, manual journals, wallet txns)
- [ ] Null/empty values stripped from responses
- [ ] Internal IDs preserved only when needed for cross-referencing

### Non-Functional

- [ ] Response size reduced by 40%+ for typical list queries (measure before/after)

### Testing

- [ ] Unit tests for each entity formatter
- [ ] Unit tests for summary statistics calculation
- [ ] Unit tests for compact mode
- [ ] Before/after size comparison test
- [ ] `npm run build && npm test` pass

### Affected Files

- `src/api/responseFormatter.ts` — New file: ResponseFormatter class
- `src/api/freeeClient.ts` — Apply formatters to API responses
- `src/index.ts` — Add `compact` parameter to list tool schemas
- `src/schemas.ts` — Update Zod schemas with `compact` parameter
- `src/types/freee.ts` — Add formatted response interfaces
- `src/__tests__/api/responseFormatter.test.ts` — New file: formatter tests

### Existing Patterns

- Follow the separation pattern: formatting logic separate from API logic
- See `src/types/freee.ts` for existing entity interfaces

### Technical Considerations

- Apply formatting at the FreeeClient level (before returning to tool handler)
- Consider a `raw: boolean` escape hatch for debugging (returns unformatted response)
- Use `JSON.stringify` replacer function for null/undefined stripping
- Measure response sizes in tests to verify 40% reduction target

### Entity Formatting Rules

| Entity | Keep | Strip |
|--------|------|-------|
| Deal | id, issue_date, type, amount, partner.name, status, details[].account_item_name | partner.id, update_date, created_at, company_id |
| Invoice | id, issue_date, invoice_number, total_amount, partner.name, payment_status | partner.id, update_date, company_id, pdf_url |
| Partner | id, name, shortcut1, long_name, contact_name | update_date, company_id, available |
| Account Item | id, name, shortcut1, group_name, tax_code | corresponding_expense/income_name if null |

## 🚧 Risks & Mitigation

| Risk | Impact | Probability | Mitigation |
|------|--------|-------------|------------|
| Stripping fields LLM actually needs | High | Medium | Start conservative, expand strip list over time |
| Breaking existing LLM workflows | High | Low | Gradual rollout, `raw` escape hatch |
| Compact mode losing needed detail | Medium | Low | Clear tool descriptions explaining compact vs full |

## 🔗 Related Work

- Benefits all existing tools and new tools from #80, #82, #83
- Related to #81 (cache) — cached responses should be pre-formatted
- Related to #80 (auto-pagination) — aggregated responses use formatted entities

### 🟡 Recommended

- [ ] Test with Claude Desktop to verify tool descriptions guide LLM effectively
- [ ] Measure API call reduction in typical analysis sessions
- [ ] Measure actual response size reduction (target: 40%+)
- [ ] Monitor LLM behavior — does it request `raw: true` frequently? (may indicate over-stripping)

### ⚠️ Manual Verification

- [ ] Verify formatted responses contain all data needed for common analysis tasks
- [ ] Test compact mode with various list sizes
- [ ] Compare LLM analysis quality before/after formatting

## 🏷️ Metadata

**Priority**: Medium
**Size**: M
**Type**: Enhancement
**Labels**: enhancement

---

🤖 AI-Optimized Issue
Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by knishioka**
## Summary

- Add `ResponseFormatter` class with entity-specific formatters that strip non-essential fields (`company_id`, `available`, `partner_id`, etc.) and null/empty values from API responses
- List endpoints now return summary statistics (`total_count`, `total_income`, `total_expense`, `date_range`) alongside formatted items
- Add `compact` parameter to 6 list tools (deals, invoices, partners, account items, sections, tags) for summary-only responses

Closes #84

## Changes

| File | Action | Purpose |
|------|--------|---------|
| `src/api/responseFormatter.ts` | New | ResponseFormatter class with formatters for all entity types |
| `src/types/freee.ts` | Modified | Added formatted response interfaces (FormattedDeal, FormattedInvoice, etc.) |
| `src/schemas.ts` | Modified | Added `compact` boolean parameter to 6 list tool schemas |
| `src/index.ts` | Modified | Integrated formatters into all list and detail tool handlers |
| `src/__tests__/api/responseFormatter.test.ts` | New | 80 unit tests covering formatters, summaries, compact mode, size reduction |

## Test plan

- [x] Unit tests for each entity formatter (deal, invoice, partner, account item, section, tag)
- [x] Unit tests for summary statistics calculation
- [x] Unit tests for compact mode (empty items with summary)
- [x] Size reduction verification tests (formatted < raw)
- [x] Edge cases (empty arrays, null values, zero amounts)
- [x] `npm run build && npm test` pass (205/205 tests)
- [x] `npm run lint` clean (no new errors)
- [x] `npm run typecheck` clean
- [ ] [manual] Test with Claude Desktop to verify tool descriptions guide LLM effectively
- [ ] [manual] Measure actual response size reduction (target: 40%+)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/api/responseFormatter.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+      a.issue_date.localeCompare(b.issue_date),
+    );
+    const summary: ListSummary = {
+      total_count: deals.length,
+      total_income: deals
+        .filter((d) => d.type === 'income')
+        .reduce((sum, d) => sum + d.amount, 0),
+      total_expense: deals
+        .filter((d) => d.type === 'expense')
+        .reduce((sum, d) => sum + d.amount, 0),
```
[the remark this answers is not in the document]
Comment (knishioka) [reply src/api/responseFormatter.ts]: Fixed in commit ff379b2. Replaced double filter+reduce with a single `reduce` that accumulates both income and expense in one pass.
